stl/inc/xutility

```diff
@@ -4043,69 +4043,106 @@ inline constexpr unreachable_sentinel_t unreachable_sentinel{};
 
-template <class _Source, class _Dest>
-struct _Ptr_copy_cat : _False_copy_cat {}; // unwrap the pointer-like type and dispatch to _Ptr_cat_helper for copy
+// Types are eligible for memcopy when they are trivially copyable.
+template <class _Ty>
+_INLINE_VAR constexpr bool
+    _Can_memcopy_elements<_Ty, enable_if_t<!is_pointer_v<_Ty>, _Ty>, _Memcopy_cat::_Copy, false> =
+        is_trivially_copyable_v<_Ty>;
 
-template <class _Source, class _Dest>
-struct _Ptr_copy_cat<_Source*, _Dest*>
-    : conditional_t<is_trivially_assignable_v<_Dest&, _Source&>,
-          _Ptr_cat_helper<remove_cv_t<_Source>, remove_cv_t<_Dest>>, _False_copy_cat> {};
+template <class _Ty>
+_INLINE_VAR constexpr bool
+    _Can_memcopy_elements<_Ty, enable_if_t<!is_pointer_v<_Ty>, _Ty>, _Memcopy_cat::_Move, false> =
+        is_trivially_copyable_v<_Ty>;
 
-template <class _Source, class _Dest>
-struct _Ptr_copy_cat<move_iterator<_Source*>, _Dest*> : _Ptr_copy_cat<_Source*, _Dest*> {};
+// Types are eligible for memcopy into uninitialized memory when they are trivial.
+template <class _Ty>
+_INLINE_VAR constexpr bool
+    _Can_memcopy_elements<_Ty, enable_if_t<!is_pointer_v<_Ty>, _Ty>, _Memcopy_cat::_Copy_uninitialized, false> =
+        is_trivial_v<_Ty>;
 
-template <class _Source, class _Dest>
-struct _Ptr_move_cat : _False_copy_cat {}; // unwrap the pointer-like type and dispatch to _Ptr_cat_helper for move
+template <class _Ty>
+_INLINE_VAR constexpr bool
+    _Can_memcopy_elements<_Ty, enable_if_t<!is_pointer_v<_Ty>, _Ty>, _Memcopy_cat::_Move_uninitialized, false> =
+        is_trivial_v<_Ty>;
 
+// Arbitrary types are not eligible
+template <class _Source, class _Dest, _Memcopy_cat _Cat>
+_INLINE_VAR constexpr bool _Can_memcopy_elements<_Source, _Dest, _Cat, false> = false;
+
+// Different types are eligible for memcopy when they are trivially copy assignable.
+template <class _Source, class _Dest, _Memcopy_cat>
+_INLINE_VAR constexpr bool _Can_assign_elements = is_trivially_assignable_v<_Dest&, _Source&>;
+
+// Different types are eligible for memmove when they are trivially copy assignable.
 template <class _Source, class _Dest>
-struct _Ptr_move_cat<_Source*, _Dest*>
-    : conditional_t<is_trivially_assignable_v<_Dest&, _Source>,
-          _Ptr_cat_helper<remove_cv_t<_Source>, remove_cv_t<_Dest>>, _False_copy_cat> {};
+_INLINE_VAR constexpr bool _Can_assign_elements<_Source, _Dest, _Memcopy_cat::_Move> =
+    is_trivially_assignable_v<_Dest&, _Source>;
 
 template <class _Source, class _Dest>
-struct _Ptr_move_cat<move_iterator<_Source*>, _Dest*> : _Ptr_move_cat<_Source*, _Dest*> {};
+_INLINE_VAR constexpr bool _Can_assign_elements<_Source, _Dest, _Memcopy_cat::_Move_uninitialized> =
+    is_trivially_assignable_v<_Dest&, _Source>;
+
+// _Memcopy_is_safe<_Iter1, _Iter2> reports whether we can activate the memcopy optimization for arbitrary iterators.
+// It ignores top-level constness on the iterators and on the elements.

```

Comment:
This is not quite correct - unlike the `equal` machinery, this cares about assigning to const elements. In general, when copy-paste-modifying code and comments, everything needs to be audited to make sure that it still applies exactly.

Reply:
I removed the `remove_const` from both the iterators as well as the elements of dest